File: frontend/src/api/data_collection.ts
import axios from '@/api'
import type { UUID } from '@/types'

// ============ 类型定义 ============

export interface Scene {
  id: string
  project_id: string
  species_id?: string
  scene_name: string
  scene_description?: string
  research_output_type?: string
  data_output_type?: string
  data_total_tb?: number
  file_size_description?: string
  data_output_description?: string
  created_at: string
}

export interface Equipment {
  id: string
  project_id: string
  scene_ids: string[]
  equipment_name: string
  equipment_type: string
  key_level: number
  procurement_method?: string
  usage_plan?: string
  unit_price?: number
  total_price: number
  supplier?: string
  is_imported: boolean
  origin_country?: string
  necessity_description?: string
  plan_usage_value?: number
  plan_usage_unit?: string
  plan_usage_description?: string
  created_at: string
}

export interface Dataset {
  id: string
  project_id: string
  data_name: string
  data_type: string
  other_data_type?: string
  data_total_tb?: number
  access_permission: string
  is_shared_with_lab: boolean
  source_equipment_ids: string[]
  scene_ids: string[]
  data_description?: string
  processing_fee: number
  compute_cycle_value?: number
  compute_cycle_unit?: string
  compute_cycle_total_days?: number
  source_cycle_months?: number
  cycle_data_gb?: number
  need_purchase: boolean
  purchase_fee: number
  created_at: string
}

export interface AIModel {
  id: string
  project_id: string
  model_name: string
  model_description?: string
  model_type: string
  model_scale?: string
  parameter_count?: string
  function_type?: string
  related_data_ids: string[]
  scene_ids: string[]
  estimated_total_fee: number
  created_at: string
}

export interface RDProject {
  id: string
  project_id: string
  rd_name: string
  rd_direction?: string
  rd_content?: string
  expected_output?: string
  estimated_fee: number
  scene_ids: string[]
  created_at: string
}

export interface Species {
  id: string
  species_code: string
  species_name: string
  category: string
  description?: string
}

export interface Dictionary {
  id: string
  dict_type: string
  dict_code: string
  dict_label: string
  dict_value?: string
}

// ============ API 函数 ============

// 场景
export const sceneApi = {
  list: (projectId: string) =>
    axios.get<Scene[]>(`/projects/${projectId}/scenes`),

  get: (id: string) =>
    axios.get<Scene>(`/scenes/${id}`),

  create: (projectId: string, data: Partial<Scene>) =>
    axios.post<Scene>(`/projects/${projectId}/scenes`, data),

  update: (id: string, data: Partial<Scene>) =>
    axios.patch<Scene>(`/scenes/${id}`, data),

  delete: (id: string) =>
    axios.delete(`/scenes/${id}`)
}

// 设备
export const equipmentApi = {
  list: (projectId: string) =>
    axios.get<Equipment[]>(`/projects/${projectId}/equipments`),

  get: (id: string) =>
    axios.get<Equipment>(`/equipments/${id}`),

  create: (projectId: string, data: Partial<Equipment>) =>
    axios.post<Equipment>(`/projects/${projectId}/equipments`, data),

  update: (id: string, data: Partial<Equipment>) =>
    axios.patch<Equipment>(`/equipments/${id}`, data),

  delete: (id: string) =>
    axios.delete(`/equipments/${id}`)
}

// 数据集
export const datasetApi = {
  list: (projectId: string) =>
    axios.get<Dataset[]>(`/projects/${projectId}/datasets`),

  get: (id: string) =>
    axios.get<Dataset>(`/datasets/${id}`),

  create: (projectId: string, data: Partial<Dataset>) =>
    axios.post<Dataset>(`/projects/${projectId}/datasets`, data),

  update: (id: string, data: Partial<Dataset>) =>
    axios.patch<Dataset>(`/datasets/${id}`, data),

  delete: (id: string) =>
    axios.delete(`/datasets/${id}`)
}

// AI模型
export const aiModelApi = {
  list: (projectId: string) =>
    axios.get<AIModel[]>(`/projects/${projectId}/ai-models`),

  get: (id: string) =>
    axios.get<AIModel>(`/ai-models/${id}`),

  create: (projectId: string, data: Partial<AIModel>) =>
    axios.post<AIModel>(`/projects/${projectId}/ai-models`, data),

  update: (id: string, data: Partial<AIModel>) =>
    axios.patch<AIModel>(`/ai-models/${id}`, data),

  delete: (id: string) =>
    axios.delete(`/ai-models/${id}`)
}

// 研发项目
export const rdProjectApi = {
  list: (projectId: string) =>
    axios.get<RDProject[]>(`/projects/${projectId}/rd-projects`),

  get: (id: string) =>
    axios.get<RDProject>(`/rd-projects/${id}`),

  create: (projectId: string, data: Partial<RDProject>) =>
    axios.post<RDProject>(`/projects/${projectId}/rd-projects`, data),

  update: (id: string, data: Partial<RDProject>) =>
    axios.patch<RDProject>(`/rd-projects/${id}`, data),

  delete: (id: string) =>
    axios.delete(`/rd-projects/${id}`)
}

// 字典
export const dictionaryApi = {
  list: (dictType?: string) =>
    axios.get<Dictionary[]>('/dictionaries', { params: { dict_type: dictType } }),

  create: (data: Partial<Dictionary>) =>
    axios.post<Dictionary>('/dictionaries', data),

  update: (id: string, data: Partial<Dictionary>) =>
    axios.patch<Dictionary>(`/dictionaries/${id}`, data),

  delete: (id: string) =>
    axios.delete(`/dictionaries/${id}`)
}

// 物种
export const speciesApi = {
  list: () =>
    axios.get<Species[]>('/species'),

  create: (data: Partial<Species>) =>
    axios.post<Species>('/species', data),

  update: (id: string, data: Partial<Species>) =>
    axios.patch<Species>(`/species/${id}`, data),

  delete: (id: string) =>
    axios.delete(`/species/${id}`)
}

File: backend/app/models/data_collection.py
from sqlalchemy import Column, String, Integer, DateTime, Boolean, ForeignKey, Text, Numeric, ARRAY
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid
from app.core.database import Base


class Species(Base):
    """物种分类表"""
    __tablename__ = "species"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    species_code = Column(String(50), unique=True, nullable=False)
    species_name = Column(String(100), nullable=False)
    parent_id = Column(UUID(as_uuid=True), ForeignKey("species.id"), nullable=True)
    category = Column(String(50), nullable=False)  # crop, horticulture, poultry, mushroom, microorganism
    description = Column(Text)
    sort_order = Column(Integer, default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    # 自关联
    parent = relationship("Species", remote_side=[id], backref="children")


class Scene(Base):
    """场景表"""
    __tablename__ = "scenes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    species_id = Column(UUID(as_uuid=True), ForeignKey("species.id"), nullable=True)
    scene_name = Column(String(255), nullable=False)
    scene_description = Column(Text)  # 场景描述（对文本生成很重要）
    research_output_type = Column(String(100))  # 科研产出（成果类型）
    research_output_data = Column(String(100))  # 科研产出-数据产出
    data_output_type = Column(String(100))  # 数据产出类型
    data_total_tb = Column(Numeric(10, 2))  # 数据总量(TB)
    file_size_description = Column(String(500))  # 文件大小描述
    data_output_description = Column(Text)  # 数据产出说明
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    deleted_at = Column(DateTime, nullable=True)

    # 关系
    project = relationship("Project", backref="scenes")
    species = relationship("Species", backref="scenes")


class Equipment(Base):
    """设备表"""
    __tablename__ = "equipments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    scene_ids = Column(ARRAY(UUID(as_uuid=True)), default=[])  # 关联场景ID数组
    equipment_name = Column(String(255), nullable=False)
    equipment_type = Column(String(50), nullable=False)  # storage, virtual, computing, sensor, robot
    key_level = Column(Integer, default=1)  # 关键星级 1-5
    procurement_method = Column(String(50))  # 采购方式
    usage_plan = Column(Text)  # 使用计划情况
    unit_price = Column(Numeric(15, 2))  # 单价（元）
    total_price = Column(Numeric(15, 2), nullable=False)  # 总价（元）
    supplier = Column(String(255))  # 供应商
    is_imported = Column(Boolean, default=False)  # 是否进口
    need_quote_seal = Column(Boolean, default=False)  # 是否需要报价盖章
    origin_country = Column(String(50))  # 国产或进口
    supplier_1 = Column(String(255))
    supplier_2 = Column(String(255))
    supplier_3 = Column(String(255))
    final_supplier = Column(String(255))
    plan_usage_value = Column(Numeric(10, 2))  # 计划使用数值
    plan_usage_unit = Column(String(20))  # 计划使用单位
    plan_usage_description = Column(Text)  # 计划使用说明
    necessity_description = Column(Text)  # 必要性与匹配性说明
    purchase_time = Column(DateTime)  # 计划购置时间 (stored as date in DB)
    commissioning_time = Column(DateTime)  # 计划投用时间 (stored as date in DB)
    data_output_type = Column(String(100))  # 数据输出类型
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    deleted_at = Column(DateTime, nullable=True)

    # 关系
    project = relationship("Project", backref="equipments")


class Dataset(Base):
    """数据集表"""
    __tablename__ = "datasets"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    data_name = Column(String(255), nullable=False)
    data_type = Column(String(50), nullable=False)  # 图像、文本、基因组、表型等
    other_data_type = Column(String(100))  # 其他数据类型
    data_total_tb = Column(Numeric(10, 2))  # 数据总量(TB)
    access_permission = Column(String(50), default='public')  # 访问权限
    is_shared_with_lab = Column(Boolean, default=False)  # 是否与实验室共享
    source_equipment_ids = Column(ARRAY(UUID(as_uuid=True)), default=[])  # 来源设备ID数组
    scene_ids = Column(ARRAY(UUID(as_uuid=True)), default=[])  # 关联场景ID数组
    data_description = Column(Text)  # 数据描述
    processing_fee = Column(Numeric(15, 2), default=0)  # 数据处理费（万元）
    compute_cycle_value = Column(Integer)  # 计算周期数值
    compute_cycle_unit = Column(String(20))  # 计算周期单位
    compute_cycle_total_days = Column(Integer)  # 总计（天）
    source_cycle_months = Column(Integer)  # 来源周期（月）
    cycle_data_gb = Column(Numeric(10, 2))  # 周期数据量(GB)
    need_purchase = Column(Boolean, default=False)  # 是否需要购买
    purchase_fee = Column(Numeric(15, 2), default=0)  # 购买费用（万元）
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    deleted_at = Column(DateTime, nullable=True)

    # 关系
    project = relationship("Project", backref="datasets")


class AIModel(Base):
    """AI模型表"""
    __tablename__ = "ai_models"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    model_name = Column(String(255), nullable=False)
    model_description = Column(Text)
    model_type = Column(String(50), nullable=False)  # deep_learning, machine_learning, llm
    model_scale = Column(String(50))  # 模型规模
    parameter_count = Column(String(50))  # 参数量
    function_type = Column(String(50))  # 功能类型：training, inference
    related_data_ids = Column(ARRAY(UUID(as_uuid=True)), default=[])  # 关联数据ID数组
    scene_ids = Column(ARRAY(UUID(as_uuid=True)), default=[])  # 关联场景ID数组
    source_equipment_ids = Column(ARRAY(UUID(as_uuid=True)), default=[])  # 关联设备ID数组
    estimated_total_fee = Column(Numeric(15, 2))  # 预计总费用（万元）
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    deleted_at = Column(DateTime, nullable=True)

    # 关系
    project = relationship("Project", backref="ai_models")


class RDProject(Base):
    """研发项目表"""
    __tablename__ = "rd_projects"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False)
    rd_name = Column(String(255), nullable=False)  # 项目名称
    rd_direction = Column(String(100))  # 研发方向
    rd_content = Column(Text)  # 研发内容
    expected_output = Column(Text)  # 预期成果
    estimated_fee = Column(Numeric(15, 2), nullable=False)  # 预估费用（万元）
    scene_ids = Column(ARRAY(UUID(as_uuid=True)), default=[])  # 关联场景ID数组
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id"))
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    deleted_at = Column(DateTime, nullable=True)

    # 关系
    project = relationship("Project", backref="rd_projects")


class Dictionary(Base):
    """字典表（用于动态扩展）"""
    __tablename__ = "dictionaries"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    dict_type = Column(String(50), nullable=False)  # equipment_type, data_type, model_type
    dict_code = Column(String(50), nullable=False)
    dict_label = Column(String(100), nullable=False)
    dict_value = Column(String(500))
    parent_id = Column(UUID(as_uuid=True), ForeignKey("dictionaries.id"), nullable=True)
    sort_order = Column(Integer, default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    # 自关联
    parent = relationship("Dictionary", remote_side=[id], backref="children")
